### ISSUE TITLE
Experiment tracking onboarding

### ISSUE THREAD

**Issue by dberenbaum** [iterative/vscode-dvc#2725]
Related to #2608, this issue is to focus on the onboarding experience described in [https://github.com/iterative/vscode-dvc/issues/2608#issuecomment-1295204054](https://github.com/iterative/vscode-dvc/issues/2608#issuecomment-1295204054).

I think a main priority should be tracking (viewing and comparing) experiments generated this way because:

- Experiment tracking is the most fully developed and obviously useful part of the extension already.
- It provides an easy onboarding experience for new users.

Note that this workflow doesn't include running/queuing experiments, so I don't think those features need to be prioritized as high. I've also found that I get more value out of the extension for viewing/comparing experiments than running/queuing them.

With that in mind, I think the current get started has too many sections (the command palette also has an overwhelming number of options). Can we refocus getting started on a workflow like:

1. Install the extension
2. Install DVC
3. Show a snippet/template/docs for how to insert dvclive into a script/notebook
4. Bring up the experiments and plots views (it’s too hard to find the experiments table and plots view as well as to see them side by side, which I find the most helpful view).

**Comment by shcheklein** [comment]
thanks @dberenbaum for the feedback.

> Experiment tracking is the most fully developed and obviously useful part of the extension already.

I'm not sure about this. It's still less stable when you actually use it (it's slow and is breaking). Table in general is the most stable component at the moment.

> 1. Install the extension

This is done. Only after that they could see the get started. We can't include this into get started _inside_ the extension. May be into a separate doc? But to be honest feels like an "obvious" step.
 
> 2. Install DVC

Already there (if we talk about get started).

> 3. Show a snippet/template/docs for how to insert dvclive into a script/notebook

Yep, we can update the page when DVCLive is ready.

> 4. Bring up the experiments and plots views (it’s too hard to find the experiments table and plots view as well as to see them side by side, which I find the most helpful view).

Duplicate #2683 to some extent. My 2cs - feel like a micro optimization. I personally use run functionality, and don't use CLI for that at all (that was the purpose of the extension tbh - so that ppl don't have to do both). It's like when you run your python code there, you don't use CLI.

---------

So, to summarize and make it a bit more actionable - I suggest to keep the "Show a snippet/template/docs for how to insert dvclive into a script/notebook" when DVCLive is ready and move it as a checkbox to the onboarding story? wdyt?

**Comment by dberenbaum** [comment]
> > Experiment tracking is the most fully developed and obviously useful part of the extension already.
> 
> I'm not sure about this. It's still less stable when you actually use it (it's slow and is breaking). Table in general is the most stable component at the moment.

Sorry, maybe "experiment tracking" was too vague. I meant the table(+plots).

> > 7\. Bring up the experiments and plots views (it’s too hard to find the experiments table and plots view as well as to see them side by side, which I find the most helpful view).
> 
> Duplicate #2683 to some extent. My 2cs - feel like a micro optimization. I personally use run functionality, and don't use CLI for that at all (that was the purpose of the extension tbh - so that ppl don't have to do both). It's like when you run your python code there, you don't use CLI.

The dvclive onboarding workflow isn't using `dvc exp run/queue`. I'm not saying we should get rid of running/queuing features, but IMHO they don't need to be as high priority, especially if the focus is onboarding.

> So, to summarize and make it a bit more actionable - I suggest to keep the "Show a snippet/template/docs for how to insert dvclive into a script/notebook" when DVCLive is ready and move it as a checkbox to the onboarding story? wdyt?

I'm fine to move this and make it a checkbox in the other story.

However, I think this misses the bigger point of how overwhelming the extension is. Is the plan to add a 12th section to get started? Can it be pared down to a specific onboarding workflow instead of showing every feature? What do you think about funneling new users to the upcoming dvclive workflow?

**Comment by shcheklein** [comment]
Okay, I see. Thanks, Dave, all good and important questions. Could you please btw point me (in Slack is fine) to the recent spec for the DVCLive integration?

I think I would need to clarify - I don't consider Get started as an onboarding experience. It only for a small fraction solves the problem, but I doubt that any changes (10 or 12 sections, DVCLive focused or not) would move the needle. (we can still optimize it and change to be new DVCLive compatible, move current things into "Pipelines" advanced section, we will see - open to discuss it).

It should be product led onboarding. Ideally, product should be onboarding into itself.

The way I see it:

- We add buttons to the panel - open plots, open experiments (not sure about split or not, but it's less important - it's an opinionated optimization)
- Buttons are always functional and enabled. If you have DVC, if you don't have it. If it's a DVC project or not.
- If there are no metrics, no DVC installed yet. We can and should utilize empty space on those webviews to explain and provide actions, snippets, etc on how to get there. It should be actionable ideally. First, we can start with a snippet, but we should have a button that generates `dvc.yaml`, generates a template of a project, etc.
- Extension also install DVC, DVCLive, etc (similar to how Jupyter extension does)

> The dvclive onboarding workflow isn't using dvc exp run/queue.

That's fine. Those buttons can stay disabled. It's similar to VS Code running python + run configurations. In this case `dvc.yaml` pipeline serves as a run config.

**Comment by dberenbaum** [comment]
> I think I would need to clarify - I don't consider Get started as an onboarding experience. It only for a small fraction solves the problem, but I doubt that any changes (10 or 12 sections, DVCLive focused or not) would move the needle. (we can still optimize it and change to be new DVCLive compatible, move current things into "Pipelines" advanced section, we will see - open to discuss it).
> 
> It should be product led onboarding. Ideally, product should be onboarding into itself.

Makes sense overall. Of course, we want the product to be self-onboarding, but if we were confident in that, I don't think these docs would be suggested as the first step for users. IMHO it would make sense to either:

1. If it's not intended to be about onboarding, change the name to "User guide," "docs," or something else besides "Get started," and don't suggest it as the first step for users.
2. If it's at least intended to be one path for onboarding, consider whether more drastic changes can be made -- not 10 vs 12 sections, but can it be 5 or fewer?

> Could you please btw point me (in Slack is fine) to the recent spec for the DVCLive integration?

It's in https://www.notion.so/iterative/Python-Notebook-Experiments-5a92a6f1677841bfb3613983f75a60b2. Let's discuss how much to lean into this scenario for VS Code. It could simplify onboarding needs if we assume new users start with this workflow since they don't need things like a template `dvc.yaml`.

**Comment by dberenbaum** [comment]
One more way of summarizing this potential onboarding workflow: The "pitch" for the DVC VS Code extension is that you can do experiment tracking locally in your IDE (obviously there's more, but I've heard you describe it in similar terms @shcheklein). IMO it would make sense to focus on this using a typical workflow of adding `log` methods to Python code and show off other features as advanced scenarios.

**Comment by dberenbaum** [comment]
Demo of the dvclive experience:

https://user-images.githubusercontent.com/2308172/204876535-0a01b1fb-28b7-48d3-87ac-11aaa7736bf5.mp4

Maybe you can have an example notebook that users can run to quickly see the table and plots? If so, it might be important to also have an easy way to drop all the experiments once users understand how it works.

**Comment by shcheklein** [comment]
This look great! :) I think we can:

- install all the deps (`dvclive`) in this case
- show a page with tabs per ML framework with code to copy paste and insert

Could you please share the `dvc.yaml` file that is generated? How do we see users transition from this to the pipelines (at least cmd + param deps) workflow in DVC? What I would have to do for that to happen?

**Comment by dberenbaum** [comment]
> * install all the deps (`dvclive`) in this case

Is the extension already ensuring DVC is installed **_as a library_**? This is also a prerequisite.

> Could you please share the `dvc.yaml` file that is generated? 

It is saved in `dvclive/dvc.yaml` (or whatever the dvclive dir is) and looks like:

```
metrics:
- metrics.json
params:
- params.yaml
plots:
- plots
```

It doesn't yet properly configure the sklearn plots (ROC, confusion matrix, etc.), which will add more plots properties.

> How do we see users transition from this to the pipelines (at least cmd + param deps) workflow in DVC? What I would have to do for that to happen?

I'm only now getting to play around with this. I'll follow up.

**Comment by dberenbaum** [comment]
You can go through the steps you normally would to set up a pipeline. The git ref generation turns off when running with `dvc exp run` or `dvc repro`.

I tried with a simple example and this is what I needed to do:
1. If you want parameter dependencies, move them into params.yaml in the root of your repo. If they are complex, you can start by copying them from dvclive/params.yaml.
2. Put code into a script. If you followed step 1, remove hard-coded parameters and corresponding `live.log_param[s]` calls, and instead read parameters into your code with `dvc.api.params_show()`.
3. Run `dvc stage add -n train -p params.yaml: -o dvclive ... python train.py`.
4. Commit your changes.
5. Start running experiments with `dvc exp run`. 

The downsides I've noted with this approach:
1. `dvc.api.params_show()` fails if `dvclive/params.yaml` doesn't exist. This should be an easy fix.
2. Parameters will show in a different column once converted into the new file. Maybe it's something we can fix in the table UI? Not a dealbreaker IMO.
3. If you manually configure your dvclive plots in your root `dvc.yaml`, you may see duplicate plots. Also not a dealbreaker and something we might be able to handle in the plots views.

**Comment by shcheklein** [comment]
Thanks, @dberenbaum !

Q: have we considered creating a regular project structure - `params.yaml` at the root, `dvc.yaml` instead of moving them into a separate `dvclive` directory? I'm not sure what is the benefit of diverging from a regular / simpler DVC project structure, esp if it complicates the transition.

C/S: `dvc.api.params_show().` needs DVC to be installed as a lib. I found it as a strong limitation last time I was trying to use it. We should move it to DVCLive I think.

C/S: can we put as comment in `dvc.yaml` a stage that would have a command to let ppl know and transition them faster. Also we can run the whole notebook as a script - it's optional to convert it I think.

**Comment by dberenbaum** [comment]
> C/S: `dvc.api.params_show().` needs DVC to be installed as a lib. I found it as a strong limitation last time I was trying to use it. We should move it to DVCLive I think.

Won't it still need dvc to be installed as a lib to collect the stage params? This is a requirement for `save_dvc_exp` inside dvclive as well.

---

> I'm not sure what is the benefit of diverging from a regular / simpler DVC project structure, esp if it complicates the transition.

The benefits are:
1. Keep all dvclive output in one place. We have found that moving any dvclive output anywhere else generates complaints and feels unexpected and messy. It also allows dvclive output to be easily copied/transferred/deleted easily.
2. Keep clean separation between dvclive config and user config. If I want to make some additional plots, either using the dvclive paths or some other data, I can add these as additional plots to my root `dvc.yaml` without worrying about messing up anything in dvclive. There's also no risk that dvclive overwrites my custom plots/metrics/params. If I change my dvclive dir, I can remove the old directory and forget about it instead of needing to remove the old paths from my root `dvc.yaml`.

I'm not sure whether combining the dvclive outputs with the overall project structure helps the transition path or adds confusion. IMHO what's logged by `log_params` is often not the same thing I would want in `params.yaml`. For example, in lightning the hyperparameters are the arguments to the LightningModule init method, for which many parameters don't make sense (like number of epochs), while others might be complex objects that are hard to list in params.yaml (model layers, optimizers, etc.).

We can keep brainstorming on ways to use this output or dvclive in general to generate starter dvc.yaml stages, but I think we need to do more than generate the dvc.yaml and should keep pipeline onboarding as a separate product goal to consider more holistically. Even if the notebook is easily converted to a dvc stage, there's minimal benefit until it is broken into a pipeline of multiple stages.

**Comment by dberenbaum** [comment]
@shcheklein Happy to close this and discuss specifics in other issues.
